### PR TITLE
[AIRFLOW-XXX] add docs showing usage of `Connection.get_uri`

### PR DESCRIPTION
**Pull request by dstandish**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: docs/howto/connection/index.rst [hunk 1]
```diff
@@ -81,6 +81,36 @@ in the scheme part of URI, so it must be changed to a hyphen character
...
+    c = Connection(
+        conn_id='some_conn',
+        conn_type='mysql',
+        host='myhost.com',
+        login='myname',
+        password='mypassword',
+        extra=json.dumps(dict(this_param='some val', that_param='other val*')),
+    )
+    print(f"AIRFLOW_CONN_{c.conn_id.upper()}='{c.get_uri()}'")
+
```
Comment (kaxil) [inline docs/howto/connection/index.rst:184]: Can you add an example output of such URI

Comment (dstandish) [reply docs/howto/connection/index.rst:184]: good idea
i have amended
thanks for the review

**Review by kaxil — APPROVED** [review]
Thanks @dstandish , LGTM.

Waiting for the CI to be green

**Comment by dstandish** [comment]
@kaxil after taking another look i could not resist doing a little refactoring of connections section

i will add to this existing PR, but in separate commit

if you think it's too much i can chop it and address later separately.  will be pushing in next few mins.

File: docs/howto/connection/index.rst [hunk 2]
```diff
@@ -81,6 +81,38 @@ in the scheme part of URI, so it must be changed to a hyphen character
...
 Query parameters are parsed to one-dimensional dict and then used to fill extra.
 
+Generating a connection URI
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Building the URI for a connection can be tricky.  To make it easier, the
+:py:class:`~airflow.models.connection.Connection` class has a convenience method
+:py:meth:`~airflow.models.connection.Connection.get_uri`.  It can be used like so:
+
+.. code-block:: python
```
Comment (mik-laj) [inline docs/howto/connection/index.rst]:
```suggestion
.. code-block:: pycon
```
Reference: https://pygments.org/docs/lexers/#pygments.lexers.python.PythonConsoleLexer

File: docs/howto/connection/index.rst [hunk 3]
```diff
@@ -61,29 +61,162 @@ to the connection you wish to edit in the connection list.
...
+        password='my-password',
+        host='my-host',
+        port=5432,
+        schema='my-schema',
+        extra=json.dumps(dict(param1='val1', param2='val2'))
+    )
+
+You can verify a URI is parsed correctly like so:
+
+.. code-block:: python
```
Comment (mik-laj) [inline docs/howto/connection/index.rst]:
```suggestion
.. code-block:: pycon
```

File: docs/howto/connection/index.rst (hunk 2 above)
Comment (dstandish) [reply docs/howto/connection/index.rst]: cool! thank you will do this

**Comment by dstandish** [comment]
OK @mik-laj i have switched to `pycon` where appropriate.  at first i thought you were pointing out a typo :)

And @kaxil  the refactor commit has been pushed.

This second commit makes the following enhancements:
* simplify description of naming convention for environment variables
* add small notes how `.bashrc` and docker `.env` are different
* add section on using the CLI (previously there was no mention of using the CLI)
* make URI discussion separate from environment variables (because URI can also be used in CLI)
* explain how to debug URI parsing, and handle special characters (though i note you can use `get_uri` and not worry about the details)

File: docs/howto/connection/index.rst [hunk 4]
```diff
@@ -61,29 +59,167 @@ to the connection you wish to edit in the connection list.
...
+    There is only one underscore surrounding ``CONN``.  This is in contrast with the way ``airflow.cfg``
+    parameters are stored, where two underscores surround the config section name.
+
+The value of this environment variable must use airflow's URI format for connections.  See the section
+:ref:`Generating a Connection URI <generating_connection_uri>` for more details.
+
+Using .bashrc (or similar)
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If storing the environment variable in something like ``~/.bashrc``, you would add it ike so:
```
Comment (kaxil) [inline docs/howto/connection/index.rst]:
```suggestion
If storing the environment variable in something like ``~/.bashrc``, add it as follow:
```

File: docs/howto/connection/index.rst [hunk 5]
```diff
@@ -61,29 +59,167 @@ to the connection you wish to edit in the connection list.
...
-must be all uppercase).
+.. note::
 
-Airflow assumes the value returned from the environment variable to be in a URI
-format (e.g. ``postgres://user:password@localhost:5432/master`` or
-``s3://accesskey:secretkey@S3``). The underscore character is not allowed
-in the scheme part of URI, so it must be changed to a hyphen character
-(e.g. ``google-compute-platform`` if ``conn_type`` is ``google_compute_platform``).
-Query parameters are parsed to one-dimensional dict and then used to fill extra.
+    The params param1 and param2 are just examples; you may supply arbitrary urlencoded json-serializable data there.
```
Comment (kaxil) [inline docs/howto/connection/index.rst]:
```suggestion
    The params ``param1`` and ``param2`` are just examples; you may supply arbitrary urlencoded json-serializable data there.
```

**Comment by dstandish** [comment]
updated; thanks

**Comment by kaxil** [comment]
Good addition, thanks @dstandish

**Pull request merged**